java/code/src/com/suse/manager/webui/services/SaltServerActionService.java

```diff
@@ -1590,6 +1606,96 @@ else if (state.equals("rebooted") && ((VirtualizationRebootAction)action).isForc
+        String initrd = dist.getInitrd();
+
+        List<String> nameParts = Arrays.asList(dist.getName().split(":"));
+        String saltFSKernel = Paths.get(nameParts.get(1), nameParts.get(0), new File(kernel).getName()).toString();
+        String saltFSInitrd = Paths.get(nameParts.get(1), nameParts.get(0), new File(initrd).getName()).toString();
+        KickstartableTree tree = KickstartFactory.lookupKickstartTreeByLabel(nameParts.get(0),
+                OrgFactory.lookupById(Long.valueOf(nameParts.get(1))));
+        tree.createOrUpdateSaltFS();
+        Map<String, Object> pillar = new HashMap<>();
+        pillar.put("uyuni-reinstall-kernel", saltFSKernel);
+        pillar.put("uyuni-reinstall-initrd", saltFSInitrd);
+        pillar.put("uyuni-reinstall-kopts", buildKernelOptions(system, profile, dist, host));
+        pillar.put("uyuni-reinstall-name", "reinstall-system");
+
+        ret.put(State.apply(Arrays.asList(KICKSTART_INITIATE), Optional.of(pillar)), minions);
+
+        return ret;
+    }
+
+    private String buildKernelOptions(SystemRecord sys, Profile prof, Distro dist, String host) {
+        String breed = dist.getBreed();
+        Map<String, Object> kopts =
+                Stream.of(dist.getKernelOptions(), prof.getKernelOptions(), sys.getKernelOptions())
+                .flatMap(map -> map.entrySet().stream())
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        Map.Entry::getValue,
+                        (v1, v2) -> v2));
+        if (breed.equals("suse")) {
+            //SUSE is not using 'text'. Instead 'textmode' is used as kernel option.
+            if (kopts.containsKey("textmode")) {
+                kopts.remove("text");
+            }
+            else if (kopts.containsKey("text")) {
+                kopts.remove("text");
+                kopts.put("textmode", "1");
+            }
+        }
+        // no additional initrd parameter allowed
+        kopts.remove("initrd");
+        String kernelOptions = convertOptionsMap(kopts);
+        String autoinst = "http://" + host + "/cblr/svc/op/autoinstall/system/" + sys.getName();
+
+        if (StringUtils.isBlank(breed) || breed.equals("redhat")) {
+           kernelOptions += " kssendmac ks=" + autoinst;
+        }
+        else if (breed.equals("suse")) {
+            kernelOptions += "autoyast=" + autoinst;
+        }
+        else if (breed.equals("debian") || breed.equals("ubuntu")) {
+            kernelOptions += "auto-install/enable=true priority=critical netcfg/choose_interface=auto url=" + autoinst;
+        }
+        return kernelOptions;
+    }
+
+    private String convertOptionsMap(Map<String, Object> map) {
+        StringBuilder string = new StringBuilder();
+        for (String key : map.keySet()) {
+            List<String> keyList;
+            try {
+                 keyList = (List)map.get(key);
+            }
+            catch (ClassCastException e) {
```

Comment:
This is a copy of a method already used in the cobbler module. Yes, it could be list or string.
I just do not want to change the code which is working since years and introduce errors because we do not know all the possible corner cases .